### PR TITLE
Add s2i-core version

### PR DESCRIPTION
**Pull request by omron93**
Discussed in https://github.com/sclorg/s2i-base-container/issues/115

image name | size | content | used by images
------------ | ------------- | ------------- | ------------- | 
s2i-core |  ~183MB | rhel7 base + s2i settings |  webstack, dbs |
s2i-base | ~423MB | s2i-core + libraries | perl python, php, ruby |

(npm not added yet)

Which rpm packages should be installed in `s2i-core` and which in `s2i-base`?

I've choosed `findutils, gettext, scl-utils, tar, yum-utils` for `core` version. (installed in by database images or used in s2i scripts)

@bparees @hhorak @jsvgoncalves @torsava @GrahamDumpleton @pkubatrh Please take a look. Which rpms to install in `s2i-core`?

### CODE REVIEW

File: base/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+# This is the list of basic dependencies that all language Docker image can
+# consume.
+# Also setup the 'openshift' user that is used for the build execution and for the
+# application runtime execution.
+# TODO: Use better UID and GID values
+RUN rpmkeys --import file:///etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7 && \
+  INSTALL_PKGS="autoconf \
+  automake \
+  bsdtar \
```
Comment (bparees) [inline base/Dockerfile]: i think you need bsdtar in core (needed for s2i).

Comment (omron93) [inline base/Dockerfile]: Thanks, added.

File: base/test/run [hunk 2]
```diff
@@ -6,7 +6,7 @@
 # IMAGE_NAME specifies a name of the candidate image used for testing.
 # The image has to be available before this script is executed.
 #
-IMAGE_NAME=${IMAGE_NAME-openshift/base-centos7-candidate}
+IMAGE_NAME=${IMAGE_NAME-centos/s2i-base-centos7-candidate}
```
Comment (torsava) [inline base/test/run]: Just out of curiosity, what's the reason behind this change?

File: base/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# This image is the base image for all OpenShift v3 language Docker images.
+FROM centos/s2i-core-centos7
+
+MAINTAINER Jakub Hadvig <jhadvig@redhat.com>
+
+# This is the list of basic dependencies that all language Docker image can
+# consume.
+# Also setup the 'openshift' user that is used for the build execution and for the
+# application runtime execution.
```
Comment (torsava) [inline base/Dockerfile]: The user is not being set in the base so the comment should be removed. Also in tho .rhel7 version

File: base/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+# TODO: Use better UID and GID values
+RUN rpmkeys --import file:///etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7 && \
```
Comment (torsava) [inline base/Dockerfile]: Can this be omitted since it's already in core?

File: base/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  patch \
+  postgresql-devel \
+  procps-ng \
+  sqlite-devel \
+  unzip \
+  wget \
+  which \
+  zlib-devel" && \
+  mkdir -p ${HOME}/.pki/nssdb && \
+  chown -R 1001:0 ${HOME}/.pki && \
```
Comment (torsava) [inline base/Dockerfile]: I've tested it and it seems these two lines above can also be safely removed from the base.

File: base/Dockerfile.rhel7 [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+      release="1" \
+      architecture="x86_64"
+
+# This is the list of basic dependencies that all language Docker image can
+# consume.
+# Also setup the 'openshift' user that is used for the build execution and for the
+# application runtime execution.
+# TODO: Use better UID and GID values
+RUN yum repolist > /dev/null && \
+  yum-config-manager --enable rhel-7-server-optional-rpms && \
```
Comment (torsava) [inline base/Dockerfile.rhel7]: These two lines above should be able to be omitted as well.

**Comment by torsava** [comment]
I've tried building [s2i-python-container](https://github.com/sclorg/s2i-python-container) on top of the new base image and everything went well, all tests are passing. But let's see how dbs do when build against the core.

Just a note: I would have appreciated splitting the change into more commits so the diffs are easier to navigate.

I've made a few comments, however, overall the PR looks good to me. :+1:

File: base/test/run (hunk 2 above)
Comment (omron93) [inline base/test/run]: This image is no longer built to openshift/ namespace. It is built as `centos/s2i-base-centos7`. This changes was forgotten to be done in #114

File: base/Dockerfile (hunk 3 above)
Comment (omron93) [inline base/Dockerfile]: Thanks, updated.

File: base/Dockerfile (hunk 4 above)
Comment (omron93) [inline base/Dockerfile]: Thanks, updated.

File: base/Dockerfile (hunk 5 above)
Comment (omron93) [inline base/Dockerfile]: Thanks, updated.

File: base/Dockerfile.rhel7 (hunk 6 above)
Comment (omron93) [inline base/Dockerfile.rhel7]: Thanks, updated.

**Comment by omron93** [comment]
@torsava Thanks for the review.

>I've tried building s2i-python-container on top of the new base image and everything went well, all tests are passing. But let's see how dbs do when build against the core.

Github CI for container images is testing changes also for all images which uses this image as a base. So if CI pass tests for current s2i based images are passing too.

I'll try to test PRs for database images manually.

> Just a note: I would have appreciated splitting the change into more commits so the diffs are easier to navigate.

Sorry. I've split it into more commits. I hope it is easier to review now.

File: base/test/run (hunk 2 above)
Comment (torsava) [inline base/test/run]: Ah, thanks!

File: core/README.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+* `PROMPT_COMMAND`: enables the collection in interactive shell
+
+Two examples:
+* If you specify `BASH_ENV`, then all your `#!/bin/bash` scripts
+do not need to call `scl enable`.
+* If you specify `PROMPT_COMMAND`, then on execution of the
+`docker exec ... /bin/bash` command, the collection will be automatically enabled.
+
+*Note*:
+The language interpreter executables in the collection (e.g., `ruby`)
```
Comment (pkubatrh) [inline core/README.md]: As this is `s2i-core` this should not mention language interpreter executables as none of the language images will be built directly upon `s2i-core` afaik.

Maybe write it in a general manner, something along the lines "Executables provided by the image layered upon `s2i-core` might not be present in a directory named in the `PATH` ..."

**Comment by torsava** [comment]
@omron93 Thanks for the commit split, it's much more readable now. And the changes look good as well!

File: core/README.md (hunk 7 above)
Comment (omron93) [inline core/README.md]: Thanks. I've made it more generic.

**Comment by omron93** [comment]
https://github.com/sclorg/mongodb-container/pull/239 and https://github.com/sclorg/mariadb-container/pull/32 are working with `centos/s2i-core-centos7` as the base image.

**Comment by pkubatrh** [comment]
LGTM, however we need to wait with merging this until after we manage to sync up in the rhel space.

**Comment by torsava** [comment]
@omron93 Regarding the movement from the `contrib` directory structure to the `root` directory structure as mandated by sclorg/container-common-scripts#8: Should I make a new PR or will you incorporate it here into yours?

**Comment by omron93** [comment]
> Should I make a new PR or will you incorporate it here into yours?

@torsava I can add the change into this PR.

**Comment by omron93** [comment]
Rebased + `contrib` moved to `/root/opt/app-root`.

**Comment by hhorak** [comment]
Sorry, I broke your PR again by merging #118. Anyway, there are two another issues, that make the tests fail:
1) Failure on this line: https://github.com/omron93/s2i-base-container/blob/core-version/core/Dockerfile#L54
```
Step 8 : COPY bin/ /usr/bin/
lstat bin/: no such file or directory
```
2) RHEL testing failing during squashing, maybe #110 might help (just re-started the test there to see whether docker-squash works)

**Comment by omron93** [comment]
Rebased.

> Failure on this line: https://github.com/omron93/s2i-base-container/blob/core-version/core/Dockerfile#L54
>Step 8 : COPY bin/ /usr/bin/
lstat bin/: no such file or directory

Thanks. It is fixed now.

> RHEL testing failing during squashing, maybe #110 might help (just re-started the test there to see whether docker-squash works)

This PR adds using of https://github.com/sclorg/container-common-scripts . So squashing should be updated there. (it needs to be fixed too... hopefully I'll do it next week)

**Comment by hhorak** [comment]
@omron93 It looks like it is a good time for merging, so once you have some time to make the final rebase, please do and merge. Review is done I guess.

**Comment by omron93** [comment]
Tests are passing. So merging.
